narwhal/network/src/client.rs

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::BTreeMap, sync::Arc, time::Duration};
+
+use anemo::{PeerId, Request};
+use async_trait::async_trait;
+use crypto::{traits::KeyPair, NetworkKeyPair, NetworkPublicKey};
+use mysten_common::sync::notify_once::NotifyOnce;
+use parking_lot::RwLock;
+use tokio::{select, time::sleep};
+use tracing::debug;
+use types::{
+    error::LocalClientError, PrimaryToWorker, WorkerOthersBatchMessage, WorkerOurBatchMessage,
+    WorkerSynchronizeMessage, WorkerToPrimary,
+};
+
+use crate::traits::{PrimaryToWorkerClient, WorkerToPrimaryClient};
+
+/// NetworkClient provides the interface to send requests to other nodes, and call other components
+/// directly if they live in the same process. It is used by both primary and worker(s).
+///
+/// Currently this only supports local direct calls, and it will be extended to support remote
+/// network calls.
+#[derive(Clone)]
+pub struct NetworkClient {
+    inner: Arc<RwLock<Inner>>,
+    shutdown_notify: Arc<NotifyOnce>,
```

Comment:
This is using a different convention than the rest of Narwhal for shutdown, and that could cause a failure to shutdown. Replacing this with the `PreSubscribedBroadcastSender` and following the convention will prevent that issue.

Reply:
Shutdown of NetworkClient is triggered conceptually outside of PrimaryNode and WorkerNode. I don't think PreSubscribedBroadcastSender would work. If consistency is the concern, I think wrapping NotifyOnce with a type that only allows wait() might be better eventually, because it does not need to specify the number of waiters upfront.

Reply:
What is the motivation for having the client created at the NarwhalManager level?

Reply:
Creating NetworkClient in NarwhalManager helps to wire primary and workers together, without using global objects.

Reply:
Btw, we should totally get rid of `PreSubscribedBroadcastSender` - I think we even discussed it in the PR that added it, that NotifyOnce in sui is what should be used to send those kind of signals

Reply:
Yes, and as I mentioned there, NotifyOnce cannot be duplicated and shared across different tasks. Can we ensure in this case that it is not possible to send a shutdown signal to the client between when the client is created and when we start listening to shutdown?

Reply:
@laura-makdah I am not sure what you talking about to be honest... We use NotifyOnce from many different tasks in sui. You can use `Arc<NotifyOnce>` if you need to clone it.

Reply:
@andll I meant that it didn't fulfill what we needed for this use case of one sender multi receiver shutdown. The tokio library most likely created different channel types for the same reason. The type `PreSubscribedBroadcastSender` is like the broadcast version of NotifyOnce. 

That said, I think we are ok with the shutdown for the client using this type.

Reply:
I am quite certain you can use NotifyOnce to shutdown multiple receivers (again, this is how we use it in sui)...
The narwhal use case is a signal to shut down components right? The way NotifyOnce can be used here, is we create Arc<NotifyOnce>, pass it to every component, and then in select they would have one of the branches doing `notify_once.wait()`. What is preventing from using it this way?

Reply:
Can you link me to these different tasks that are waiting on the same instance of NotifyOnce? I see an instance called "epoch_alive_notify" but it is only wait()-ed on in one place. Which instance are you referring to.

Reply:
Yeah, I maintain that NotifyOnce is like a safe watch channel, while PreSubscribedBroadcastSender is like a safe broadcast channel.

Reply:
I think for Narwhal's use case, `NotifyOnce` is sufficient since only one shutdown notification is needed. We can look into this later if we are refactoring or fixing issues with the shutdown logic.